linkml/generators/pydanticgen/array.py

```diff
@@ -0,0 +1,396 @@
+            with_inner_union (bool): If ``True`` , the innermost type is a ``Union`` of the ``AnyShapeArray`` class
+                and ``dtype`` (default: ``False`` )
+
+        """
+        if self.dtype in ("Any", "AnyType"):
+            annotation = "AnyShapeArray"
+        else:
+            annotation = f"AnyShapeArray[{self.dtype}]"
+
+        if with_inner_union:
+            annotation = f"Union[{annotation}, {self.dtype}]"
+        return SlotResult(annotation=annotation, injected_classes=_AnyShapeArrayInjects, imports=_AnyShapeArrayImports)
+
+    def anonymous_shape(self, array: ArrayExpression) -> SlotResult:
+        if array.exact_number_dimensions:
+            return SlotResult(annotation=self._list_of_lists(array.exact_number_dimensions, self.dtype))
+        elif not array.maximum_number_dimensions and (
+            array.minimum_number_dimensions is None or array.minimum_number_dimensions == 1
+        ):
+            return self.any_shape()
+        elif array.maximum_number_dimensions:
+            min_dims = array.minimum_number_dimensions if array.minimum_number_dimensions is not None else 1
+            annotations = [
+                self._list_of_lists(i, self.dtype) for i in range(min_dims, array.maximum_number_dimensions + 1)
+            ]
+            # TODO: Format this nicely!
+            return SlotResult(annotation="Union[" + ", ".join(annotations) + "]")
+        else:
+            return SlotResult(
+                annotation=self._list_of_lists(array.minimum_number_dimensions - 1, self.any_shape().annotation),
+                injected_classes=_AnyShapeArrayInjects,
+                imports=_AnyShapeArrayImports,
+            )
+
+    def labeled_shape(self, array: ArrayExpression) -> SlotResult:
+        """
+        Constrained shapes using :func:`pydantic.conlist`
+
+        TODO:
+        - preservation of aliases
+        - (what other metadata is allowable on labeled dimensions?)
+        """
+        # generate dimensions from inside out and then format
+        range = self.dtype
+        for dimension in reversed(array.dimensions):
+            range = self._labeled_dimension(dimension, range).annotation
+
+        return SlotResult(annotation=range, imports=_ConListImports)
+
+    def mixed_shape(self, array: ArrayExpression) -> SlotResult:
+        """
+        Mixture of labeled dimensions with a max or min (or both) shape for anonymous dimensions
+        """
+        if array.exact_number_dimensions is not None:
+            if array.exact_number_dimensions > len(array.dimensions):
+                res = SlotResult(
+                    annotation=self._list_of_lists(array.exact_number_dimensions - len(array.dimensions), self.dtype)
+                )
+            elif array.exact_number_dimensions == len(array.dimensions):
+                # equivalent to labeled shape
+                return self.labeled_shape(array)
```

Comment:
```suggestion
        if array.exact_number_dimensions or (
            array.minimum_number_dimensions and array.maximum_number_dimensions and 
            array.minimum_number_dimensions == array.maximum_number_dimensions
        ):
            exact_dims = array.exact_number_dimensions or array.minimum_number_dimensions
            if exact_dims > len(array.dimensions):
                res = SlotResult(
                    annotation=self._list_of_lists(exact_dims - len(array.dimensions), self.dtype)
                )
            elif exact_dims == len(array.dimensions):
                # equivalent to labeled shape
                return self.labeled_shape(array)
```

Again to combine the cases where exact dims are provided and where min and max are provided and are equal.